src/main/java/com/bobocode/blyznytsia/bibernate/session/SessionFactoryImpl.java

```diff
@@ -0,0 +1,109 @@
+package com.bobocode.blyznytsia.bibernate.session;
+
+import com.bobocode.blyznytsia.bibernate.transaction.Transaction;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import javax.sql.DataSource;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class SessionFactoryImpl implements SessionFactory {
+
+  private final DataSource dataSource;
+  private final Set<Session> sessions;
+  private boolean closed = false;
+
+  public SessionFactoryImpl(DataSource dataSource) {
+    log.debug("Initializing SessionFactory");
+    Objects.requireNonNull(dataSource, "DataSource must not be null");
+    this.dataSource = dataSource;
+    this.sessions = ConcurrentHashMap.newKeySet();
+    log.debug("SessionFactory initialized");
+  }
+
+  // TODO VS: Change to real implementation
+  @Override
+  public Session openSession() {
+    log.trace("Opening the new session");
+    if (!isOpen()) {
+      throw new IllegalStateException("SessionFactory is closed");
+    }
+    Session newSession = new SessionImpl(this, dataSource, this::removeSession);
+    sessions.add(newSession);
+    return newSession;
+  }
+
+  @Override
+  public boolean isOpen() {
+    return !closed;
+  }
+
+  @Override
+  public void close() {
+    log.trace("Closing SessionFactory");
+    log.trace("Closing Sessions which are opened");
+    sessions.forEach(Session::close);
+    sessions.clear();
+    // TODO VS: close dataSource
```

Comment:
if close datasource will be implemented on Session level than why do we need this //todo here?

Reply:
I think that `DataSource` should be closed by `SessionFactory`, as after closing one session we can create another session with the same `DataSource`. `DataSource` is heavy object to create.

Reply:
Ok. It makes sense. But I still don't understand why we don't close it here and have this //todo instead?

maybe it should be something like:
```
   var unwrap = dataSource.unwrap(HikariDataSource.class);
   unwrap.close();
```

instead of this //todo?

Reply:
Makes sense, will do this